src/backends/perf_pt/decode.c

```diff
@@ -195,41 +196,63 @@ perf_pt_next_block(struct pt_block_decoder *decoder, int *decoder_status,
     struct pt_block block;
-    *decoder_status = pt_blk_next(decoder, &block, sizeof(block));
-    // Other +ve decoder status codes can arise here. We ignore them for now,
-    // and let them be detected by handle_events() above when we are next
-    // called.
-    if (*decoder_status < 0) {
-        if (*decoder_status == -pte_eos) {
-            // End of stream is flagged as an error in the case of pt_blk_next().
+    block.iclass = ptic_other;
+    bool first_block = true;
+    while (!block_is_terminated(&block)) {
+        if (handle_events(decoder, decoder_status, err) != true) {
+            // handle_events will have already called perf_pt_set_err().
+            return false;
+        } else if (*decoder_status & pts_eos) {
+            // End of stream.
             *addr = 0;
             return true;
-        } else {
-            // A real error.
-            perf_pt_set_err(err, perf_pt_cerror_ipt, -*decoder_status);
-            return false;
         }
-    }
-    // It's possible at this point that we get notified of an event in the
-    // stream. This will be handled in the next call to `perf_pt_next_block`.
-    if ((*decoder_status != 0) && (*decoder_status != pts_event_pending)) {
-        panic("Unexpected decoder status: %d", *decoder_status);
-    }
+        // It's possible at this point that we get notified of an event in the
+        // stream. This will be handled in the next call to `perf_pt_next_block`.
+        if ((*decoder_status != 0) && (*decoder_status != pts_event_pending)) {
+            panic("Unexpected decoder status: %d", *decoder_status);
+        }
 
-    // XXX A truncated block occurs when a block straddles a section boundary.
-    // In this case we may need some extra logic, but this should be rare.
-    if (block.truncated != 0) {
-        panic("Truncated blocks are not implemented");
-    }
+        // Now fetch the block information.
+        *decoder_status = pt_blk_next(decoder, &block, sizeof(block));
+        // Other +ve decoder status codes can arise here. We ignore them for now,
+        // and let them be detected by handle_events() above when we are next
+        // called.
+        if (*decoder_status < 0) {
+            if (*decoder_status == -pte_eos) {
+                // End of stream is flagged as an error in the case of pt_blk_next().
+                *addr = 0;
+                return true;
+            } else {
+                // A real error.
+                perf_pt_set_err(err, perf_pt_cerror_ipt, -*decoder_status);
+                return false;
+            }
+        }
+
+        // XXX A truncated block occurs when a block straddles a section boundary.
+        // In this case we may need some extra logic, but this should be rare.
+        if (block.truncated != 0) {
+            panic("Truncated blocks are not implemented");
+        }
+
+        // A block should have at least one instruction.
```

Comment:
Is this something we need to check for here?

Reply:
I'm not sure what you mean.

Reply:
Why are we checking this?

Reply:
I think it can't happen, but I want to know if it can.

Reply:
And if it happened... why would this be the right place to catch it?

Reply:
If I moved it outside the loop, I'd only be checking blocks which are the start of a fully-terminated block, and not all blocks (partial or otherwise).

Is that what you mean? I'm not 100% sure what you are asking :P

Reply:
I understand why some parts of the code might need to check this invariant holds, but this doesn't seem like the right place for it to be checked (to me, at least).

Reply:
Where would you put it?

Reply:
I don't know. It feels like it's a concern of the user of this code to me, but I could be persuaded otherwise.

Reply:
The user won't see every block libipt generates.

Imagine we get two blocks from libipt:
```
[block]
mov a, b
add b, c
[block]
jmp 0xdeadbeef
```

The first block is partial, so users of our API won't see the second block. We conceptually merge the blocks.

Now imagine the second was an empty block. I'd like to know -- as the developer of this library -- if that happened.

Get it?

Reply:
Ah, I see, because it would mean you'd have to think about merging in a different way?

Reply:
Hrm, not sure what you mean.

Basically the user doesn't care to see partial blocks, but I do care to know if a partial block (or a complete block for that matter) is empty. I don't yet know under what circumstances that can happen so I'd like for the program to crash for now.

If at the libipt level, we had:
```
[block]
mov a, b
add b, c
[block]
<empty!>
```

then the user of our API sees:
```
[block]
mov a, b
add b, c
```

They would never see the empty block, so we need to check at the lower level.